pkg/controllers/metrics/node/counter/counter.go

```diff
@@ -0,0 +1,165 @@
+			metricConditionMemoryPressureLabel,
+			metricConditionPIDPressureLabel,
+			metricConditionReadyLabel,
+			metricInstanceTypeLabel,
+			metricTopologyRegionLabel,
+			metricTopologyZoneLabel,
+			metricOsLabel,
+		},
+	)
+
+	prometheusLabelsFor = make(map[types.NamespacedName]prometheus.Labels)
+
+	conditionTypeToMetricLabel = map[v1.NodeConditionType]string{
+		v1.NodeDiskPressure:   metricConditionDiskPressureLabel,
+		v1.NodeMemoryPressure: metricConditionMemoryPressureLabel,
+		v1.NodePIDPressure:    metricConditionPIDPressureLabel,
+		v1.NodeReady:          metricConditionReadyLabel,
+	}
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(nodeCountGaugeVec)
+}
+
+// UpdateCount updates the emitted metric based on the node's current status relative to the
+// past status. If the data for `node` cannot be populated then `nil` should be passed as the
+// argument.
+func UpdateCount(ctx context.Context, name types.NamespacedName, node *v1.Node) {
+	currLabels := getLabels(node)
+	pastLabels, isKnown := prometheusLabelsFor[name]
+	switch {
+	case !isKnown && node != nil:
+		handleCreatedNode(ctx, name, currLabels)
+	case isKnown && node == nil:
+		handleDeletedNode(ctx, name, pastLabels)
+	case isKnown && node != nil:
+		handleUpdatedNode(ctx, name, currLabels, pastLabels)
+	default: // An unknown node was deleted.
+	}
+}
+
+func handleCreatedNode(ctx context.Context, name types.NamespacedName, labels prometheus.Labels) {
+	prometheusLabelsFor[name] = labels
+
+	if err := incrementNodeCount(labels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to update count metric for new node [labels=%s]: error=%s", labels, err.Error())
+	}
+}
+
+func handleDeletedNode(ctx context.Context, name types.NamespacedName, labels prometheus.Labels) {
+	delete(prometheusLabelsFor, name)
+
+	if err := decrementNodeCount(labels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to update count metric for deleted node [labels=%s]: error=%s", labels, err.Error())
+	}
+}
+
+func handleUpdatedNode(ctx context.Context, name types.NamespacedName, currLabels prometheus.Labels, pastLabels prometheus.Labels) {
+	// Only report node updates that affect tracked dimensions.
+	if reflect.DeepEqual(currLabels, pastLabels) {
+		return
+	}
+
+	prometheusLabelsFor[name] = currLabels
+
+	if err := decrementNodeCount(pastLabels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to decrement previous count for updated node [labels=%s]: error=%s", pastLabels, err.Error())
+	}
+	if err := incrementNodeCount(currLabels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to increment current count for updated node [labels=%s]: error=%s", currLabels, err.Error())
+	}
+}
+
+func getLabels(node *v1.Node) prometheus.Labels {
+	labels := make(prometheus.Labels)
+	if node == nil {
+		return labels
+	}
+
+	labels[metrics.ProvisionerLabel] = node.Labels["karpenter.sh/provisioner-name"]
+	labels[metricArchLabel] = node.Labels["kubernetes.io/arch"]
```

Comment:
you could switch out these hardcoded labels to these: https://github.com/kubernetes/api/blob/master/core/v1/well_known_labels.go

Reply:
Nice 👍